### DISCUSSION TITLE
Can't get example running!! Pico.

### DISCUSSION THREAD

**Discussion by phuzzyday** [n-elia/MAX30102-MicroPython-driver#20] [Help]
Once I finally figured out all the ways I was not following the instructions, I became sure something was wrong! I am just starting out with micropython, so please bear with me.)

I am trying to follow this - 
"To run the example in ./example folder, copy max30102/circular_buffer.py and max30102/__init__.py into the ./example/lib directory. Then, upload the ./example directory content into your microcontroller. After the upload, press the reset button of your board are you're good to go."

I set up the lib directory with the two files mentioned in the docs, and the two files in the root directory. (Boot/Main.) You talk about a reset button, there isn't one. In thonny, I had to hit CTRL-D. Is that worth having in the docs?
Anyway, it kept telling me "ImportError: no module named 'max30102'".
I opened the readme.txt from the example folder, and it told me the file name should be max30102! Not '__init__.py'!
I changed the name of that file, and am left with a new error - ;

Traceback (most recent call last):

ImportError: no module named 'network'
  File "main.py", line 6, in <module>
ImportError: can't import name MAX30102"

So my device is a pico without the network, I imagine that explains that error, but it should still run, I would think? I have a network one here, I can switch I guess...

OH! I also got another error at the start.....  from boot.py. It was unable to import 'sleep' from machine. I cut the word out to see if it would go past it, and it did.  Mind you, that was while I was starting the program with the play button in thonny instead of the CTRL-D, so that may have been it.

My wife has a serious type of sleep apnea, and I really want to log her Pulse-Ox. (She is using a CPAP also.) Thank you for the work done, can you help this newbie?

Phil

**Comment by n-elia** [comment]
Hi Phil,

first of all - for the sake of clarity - let me remind you two important considerations:
1. As the disclaimer clearly states, this work is not intended to be used in professional environments, and there are no guarantees on its functionalities. Please do not rely on it for serious medical purposes!
2. This is a driver, and therefore you will be able to get RAW data from the pulse-oximeter sensor. Then, you will have to write your own code to be fed up with RAW data and to provide as output an estimate heartrate and SpO2.

I understand that you're using Thonny IDE. The main basic thing to understand is that you have to distinguish between the filesystem of your PC and the filesystem of the Pico board. You have to prepare your folder structure on the PC as described [here](https://github.com/n-elia/MAX30102-MicroPython-driver#1b---manual-way-no-internet-access-required) (**there was an error inside the README that I just fixed**), and then you have to upload all of that on the Pico board. Once that those files are properly pasted into the Pico, whenever you reboot it, it will at first execute `boot.py` and then `main.py`.

Ctrl-D issues a soft reboot on your board. Most of the times this is enough to properly run the files. FYI, Pico board can be reset by connecting the "RUN" pin with the ground ([example](https://www.raspberrypi-spy.co.uk/2022/12/how-to-add-a-raspberry-pi-pico-reset-button/)).
Note that you do not have to press the green "play" button on Thonny, because you only need to reset the board (using Ctrl-D) and connect the terminal to see the output.

Please feel free to write any doubt, bye!

Edit: an example of what you should see on Thonny:
![image](https://user-images.githubusercontent.com/62206011/218457054-20f0f747-bcac-4e69-ad0f-88a40456df0c.png)

**Reply by skely** [reply]
Hi, there is this "file structure" error still in pypi documentation. (https://pypi.org/project/micropython-max30102/).

**Reply by n-elia** [reply]
Thanks for reporting! I will release a new version to update it ASAP 👍

**Reply by n-elia** [reply]
@skely Thanks again for your feedback, the new release is rolling out right now, and the PyPi docs will be updated accordingly! :+1:

**Comment by phuzzyday** [comment]
Thank you for the clarifications regarding actual medical use, and the actual driver functions. I was aware of those. I am not new to microcontrollers, but Python is the new part!
Grounding the reset pin just makes thonny disconnect, I found that first!
I thought after following the instructions that the two .py files in the lib folder were NOT in a max30102 folder! I have remedied that, and the code is starting properly! I'll poke around and see how I got that impression. I was trying to be very accurate with my reading.

Sadly, I am getting 'Sensor Not Found.'
I am attaching a picture of my setup, can you see where I am going wrong?

I really appreciate your help.

![20230213_124141](https://user-images.githubusercontent.com/19560393/218559173-23f94498-1558-416f-9c7e-04ceeee9e445.jpg)

**Reply by phuzzyday** [reply]
I think I know where I went wrong...

See this text?
"To directly include the library into a MicroPython project, it's sufficient to copy max30102/circular_buffer.py and max30102/__init__.py next to your main.py file, into a lib directory."

Clearly I thought the image below it was referring to something else, because I completely missed it, but here, I was thinking a bit like a computer. When you said " max30102/circular_buffer.py" I thought you were just telling me where the file was! Not to copy the directory name as well. Perhaps I am not used to the typical wording conventions.

**Reply by n-elia** [reply]
It was my fault! I forgot to update that part of the README...now I should have fixed it completely with the last commit! Thanks for making this issue pop out!!

You're right. What happens is: Micropython looks for libraries to be imported inside the `lib `directory. Our module consists of a folder, so we have to put the whole folder inside `lib`!

I hope that you can proceed now. Please, feel free to share your code with us if you manage to live estimate SpO2 and HR! And, if something strange happens with SpO2, check [this](https://github.com/n-elia/MAX30102-MicroPython-driver#sensor-clones), it may be your issue!

Bye,
Nicola

**Reply by phuzzyday** [reply]
Thank you for your help! Did you see my question about the 'sensor not found' issue and the photo? Maybe you fixed something. I'll try a new download.

EDIT: Same result. Sensor not found. Tried two sensors too! I might try an ESP32 instead. There is a minor typo in your readme file fix also..  Just so you know..

**Reply by n-elia** [reply]
Sure, let me know which is the typo and i will correct asap!

If you get that message it means that the i2c interface you're using can't find the sensor. Check carefully:
1. If the pins that you selected to initialize the i2c class are correct. You can use the pins 16, 17
2. If you connected SDA to SCL and viceversa. For i2c to work, you have to connect SDA with SDA and SLC with SCL.
3. If the sensor's board is receiving the correct voltage.

Edit: we have [some](https://github.com/n-elia/MAX30102-MicroPython-driver/discussions/5#discussioncomment-1017043) evidences that the library works on Raspberry Pi Pico, so it must be something related to i2c setup or the wiring. You can also feed the sensor with 5V Vin, if you want to try a different power source.

**Reply by phuzzyday** [reply]
Ok - I'll keep at it. The typo was in the text file you fixed in example/lib/readme.
"To manually install the library, copy the files max30102/circular_buffer.py, max30102/max30102.py and past them here."

Thanks for all the help!

**Reply by phuzzyday** [reply]
I'm a dope...  I hadn't set the pins in main.py.

**Reply by n-elia** [reply]
Hahaha it happens. This is why I always start from basics when troubleshooting! I hope it works well now, and I look forward to seeing your results!
